rollouts/controllers/rollout_controller.go

```diff
@@ -433,6 +427,65 @@ func (r *RolloutReconciler) computeTargets(ctx context.Context,
 	return targets, nil
 }
 
+func (r *RolloutReconciler) getWaveTargets(ctx context.Context, rollout *gitopsv1alpha1.Rollout, allTargets *Targets,
+	allWaves []gitopsv1alpha1.Wave) ([]WaveTarget, error) {
+	allWaveTargets := []WaveTarget{}
+
+	clusterNameToWaveTarget := make(map[string]*WaveTarget)
+
+	for i := range allWaves {
+		wave := allWaves[i]
+		thisWaveTarget := WaveTarget{Wave: &wave, Targets: &Targets{}}
+
+		waveClusters, err := r.store.ListClusters(ctx, rollout.Spec.Targets.Selector, wave.Targets.Selector)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, cluster := range waveClusters.Items {
+			clusterNameToWaveTarget[cluster.Name] = &thisWaveTarget
+		}
+
+		allWaveTargets = append(allWaveTargets, thisWaveTarget)
+	}
+
+	for _, toCreate := range allTargets.ToBeCreated {
+		wavetTargets := clusterNameToWaveTarget[toCreate.cluster.Name].Targets
+		wavetTargets.ToBeCreated = append(wavetTargets.ToBeCreated, toCreate)
+	}
+
+	for _, rrs := range allTargets.ToBeUpdated {
+		wavetTargets := clusterNameToWaveTarget[rrs.Spec.ClusterRef.Name].Targets
+		wavetTargets.ToBeUpdated = append(wavetTargets.ToBeUpdated, rrs)
+	}
+
+	for _, rrs := range allTargets.Unchanged {
+		wavetTargets := clusterNameToWaveTarget[rrs.Spec.ClusterRef.Name].Targets
+		wavetTargets.Unchanged = append(wavetTargets.Unchanged, rrs)
+	}
+
+	for _, rrs := range allTargets.ToBeDeleted {
+		// The remote root sync will be associated back to it's previous wave and then removed as part
+		// of that wave. If the previous wave the remote root sync cannot be determined, then the remote
+		// root sync will be removed with the last wave of the rollout.
+
+		waveName, found := findWaveNameForCluster(rollout, rrs.Spec.ClusterRef.Name)
+
+		if !found {
+			waveName = allWaveTargets[len(allWaveTargets)-1].Wave.Name
```

Comment:
You must have thought about it already, I was wondering about the edge case where only one wave is involved.